File: crates/hyprduck-engine-types/src/agent_chat.rs
use serde::{Deserialize, Serialize};

use hyprduck_knowledge::AnswerResponse;

use crate::{
    BrainReadScope, ContextPackEvidenceV1, ContextPackRetrievalTraceV1, SourceId,
};

pub const AGENT_CHAT_SCHEMA_VERSION: &str = "hyprduck.agent_chat.v1";

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AgentChatScopeMode {
    Auto,
    AllDocs,
    SelectedSource,
    GraphContext,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AgentChatAnswerMode {
    General,
    Evidence,
    Blocked,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AgentChatStreamStatus {
    ResolvingScope,
    RetrievingContext,
    ClassifyingQuestion,
    ConnectingProvider,
    Generating,
    ValidatingCitations,
    Complete,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AgentChatMessageRole {
    User,
    Assistant,
    System,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentChatMessage {
    pub id: String,
    pub role: AgentChatMessageRole,
    pub text: String,
    pub created_at: u64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentChatAskRequest {
    pub schema_version: String,
    pub conversation_id: String,
    #[serde(default)]
    pub assistant_message_id: Option<String>,
    pub scope: BrainReadScope,
    pub mode: AgentChatScopeMode,
    #[serde(default)]
    pub selected_node_id: Option<String>,
    #[serde(default)]
    pub source_ids: Vec<SourceId>,
    pub question: String,
    #[serde(default)]
    pub history: Vec<AgentChatMessage>,
    #[serde(default)]
    pub budget: Option<usize>,
    #[serde(default)]
    pub persist_context_pack: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentChatProviderSummary {
    pub id: String,
    pub label: String,
    pub model_id: String,
    pub hosted: bool,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentChatAskResponseData {
    pub schema_version: String,
    pub conversation_id: String,
    pub answer_mode: AgentChatAnswerMode,
    pub assistant_message: AgentChatMessage,
    pub answer: AnswerResponse,
    pub context_pack_id: String,
    #[serde(default)]
    pub persisted_context_pack_path: Option<String>,
    #[serde(default)]
    pub citations: Vec<ContextPackEvidenceV1>,
    pub retrieval_trace: ContextPackRetrievalTraceV1,
    pub provider: AgentChatProviderSummary,
    #[serde(default)]
    pub warnings: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum AgentChatStreamEvent {
    Started {
        conversation_id: String,
        assistant_message_id: String,
        provider: AgentChatProviderSummary,
        #[serde(default)]
        answer_mode: Option<AgentChatAnswerMode>,
    },
    Status {
        status: AgentChatStreamStatus,
        message: String,
    },
    Delta {
        text: String,
    },
    CitationUpdate {
        citations: Vec<ContextPackEvidenceV1>,
    },
    Final {
        result: AgentChatAskResponseData,
    },
    Error {
        code: String,
        message: String,
    },
    Stopped {
        partial_text: String,
    },
}

File: apps/desktop/src/webPreview/handlers/agentChat.ts
import type {
  AgentChatAskPayload,
  AgentChatAskResult,
  AgentChatStreamEvent,
} from "@/appTypes";

import { getWebWorkspaceFromSnapshot } from "./graph";
import {
  emitWebAgentChatEvent,
  webMockAgentChatTimers,
  webMockConfig,
} from "../state";

function createWebAgentChatResult(request: AgentChatAskPayload): AgentChatAskResult {
  const workspace = getWebWorkspaceFromSnapshot();
  if (!workspace.project) {
    throw new Error("No workspace available in preview mode.");
  }
  const answer =
    workspace.project.answerByNodeId["source:preview"] ??
    Object.values(workspace.project.answerByNodeId)[0];
  if (!answer) {
    throw new Error("No answer available for this workspace in preview mode.");
  }
  const now = Math.floor(Date.now() / 1000);
  return {
    schemaVersion: "hyprduck.agent_chat.v1",
    conversationId: request.conversationId,
    answerMode: "evidence",
    assistantMessage: {
      id: request.assistantMessageId ?? `preview-msg-${now}`,
      role: "assistant",
      text:
        answer.text ??
        "This preview answer is grounded in the sample workspace evidence.",
      createdAt: now,
    },
    answer,
    contextPackId: "ctx_web_preview",
    persistedContextPackPath: null,
    citations: answer.citations.map((citation) => ({
      evidenceRef: citation.id,
      sourceId: citation.sourceId ?? "preview",
      page: citation.pageIndex ?? 1,
      region: null,
      span: null,
      quotedText: citation.snippet,
      parseConfidence: "high",
      selectionReason: citation.provenance ?? "preview evidence",
      contentHash: "web-preview",
      evidenceType: "text",
    })),
    retrievalTrace: {
      strategy: "web_preview",
      chunksConsidered: answer.citations.length,
      chunksSelected: answer.citations.length,
      budgetRequested: request.budget ?? 8000,
      budgetUsed: answer.citations.length * 64,
      evidenceTypeTrace: {
        considered: { text: answer.citations.length },
        selected: { text: answer.citations.length },
      },
    },
    provider: {
      id: webMockConfig.provider,
      label: webMockConfig.provider === "open_router" ? "OpenRouter" : "Ollama",
      modelId: webMockConfig.model_id,
      hosted: webMockConfig.provider === "open_router",
    },
    warnings: [],
  };
}

export const agentChatHandlers = {
  agent_chat_ask: (args: { request: AgentChatAskPayload }) =>
    createWebAgentChatResult(args.request),
  agent_chat_start: (args: { request: AgentChatAskPayload }) => {
    const requestId = `preview-agent-${Date.now()}`;
    const conversationId = args.request.conversationId;
    const assistantMessageId =
      args.request.assistantMessageId ?? `preview-msg-${Date.now()}`;
    const request = { ...args.request, assistantMessageId };
    const result = createWebAgentChatResult(request);
    const text = result.assistantMessage.text || result.answer.text || "";
    const chunks = text.match(/.{1,32}/g) ?? [text];
    const timers: ReturnType<typeof setTimeout>[] = [];
    const schedule = (delay: number, event: AgentChatStreamEvent) => {
      timers.push(setTimeout(() => emitWebAgentChatEvent(event), delay));
    };
    schedule(10, {
      requestId,
      type: "started",
      conversationId,
      assistantMessageId,
      provider: result.provider,
      answerMode: result.answerMode,
    });
    schedule(40, {
      requestId,
      type: "status",
      status: "retrieving_context",
      message: "Retrieving context...",
    });
    schedule(90, {
      requestId,
      type: "status",
      status: "generating",
      message: "Generating...",
    });
    chunks.forEach((chunk, index) => {
      schedule(130 + index * 45, {
        requestId,
        type: "delta",
        text: chunk,
      });
    });
    const finalDelay = 160 + chunks.length * 45;
    schedule(finalDelay, {
      requestId,
      type: "status",
      status: "validating_citations",
      message: "Validating citations...",
    });
    schedule(finalDelay + 40, {
      requestId,
      type: "final",
      result,
    });
    timers.push(
      setTimeout(() => {
        webMockAgentChatTimers.delete(requestId);
      }, finalDelay + 60),
    );
    webMockAgentChatTimers.set(requestId, timers);
    return { requestId, conversationId, assistantMessageId };
  },
  agent_chat_stop: (args: { requestId: string }) => {
    const timers = webMockAgentChatTimers.get(args.requestId) ?? [];
    for (const timer of timers) {
      clearTimeout(timer);
    }
    webMockAgentChatTimers.delete(args.requestId);
    emitWebAgentChatEvent({
      requestId: args.requestId,
      type: "stopped",
      partialText: "",
    });
    return { stopped: timers.length > 0 };
  },
};
